pkg/apis/pipeline/v1alpha1/gcs_resource.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2018 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"flag"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	gsutilImage              = flag.String("gsutil-image", "override-with-gsutil-image:latest", "The container image containing gsutil")
+	gcsSecretVolumeMountPath = "/var/secret"
+)
+
+// GCSResource is a GCS endpoint from which to get artifacts which is required
+// by a Build/Task for context (e.g. a archive from which to build an image).
+type GCSResource struct {
+	Name           string               `json:"name"`
+	Type           PipelineResourceType `json:"type"`
+	Location       string               `json:"location"`
+	TypeDir        bool                 `json:"typeDir"`
+	DestinationDir string               `json:"destinationDir"`
+	//Secret holds a struct to indicate a field name and corresponding secret name to populate it
+	Secrets []SecretParam `json:"secrets"`
+}
+
+// NewGCSResource creates a new GCS resource to pass to knative build
+func NewGCSResource(r *PipelineResource) (*GCSResource, error) {
+	if r.Spec.Type != PipelineResourceTypeStorage {
+		return nil, fmt.Errorf("GCSResource: Cannot create a GCS resource from a %s Pipeline Resource", r.Spec.Type)
+	}
+	var location string
+	var locationSpecified, dir bool
+
+	for _, param := range r.Spec.Params {
+		switch {
+		case strings.EqualFold(param.Name, "Location"):
+			location = param.Value
+			if param.Value != "" {
+				locationSpecified = true
+			}
+		case strings.EqualFold(param.Name, "Dir"):
+			dir = true // if dir flag is present then its a dir
+		}
+	}
+
+	if !locationSpecified {
+		return nil, fmt.Errorf("GCSResource: Need Location to be specified in order to create GCS resource %s", r.Name)
+	}
+	return &GCSResource{
+		Name:     r.Name,
+		Type:     r.Spec.Type,
+		Location: location,
+		TypeDir:  dir,
+		Secrets:  r.Spec.SecretParams,
+	}, nil
+}
+
+// GetName returns the name of the resource
+func (s GCSResource) GetName() string {
+	return s.Name
+}
+
+// GetType returns the type of the resource, in this case "storage"
+func (s GCSResource) GetType() PipelineResourceType {
+	return PipelineResourceTypeStorage
+}
+
+// GetParams get params
+func (s *GCSResource) GetParams() []Param { return []Param{} }
```

Comment:
should this be returning `Location` and `Dir`?

Reply:
I looked at git, image, cluster resource as example and all of them return empty params but I can update this to return `Location`. `Dir` is like runtime param I am not sure it fits the case here.

Reply:
kk!